Tray D-Bus Menu

FlexW opened the pull request:
This pull request implements D-Bus menus for the tray icons. @ianyfan @nmschulte It seems like #5161 is dead, and I would like to take over. Please tell me if this is not the case, then I will close this pull request.

This pull request is based on #5161 and is a draft for now. Please don't test it yet. When the pull request is ready for testing, I will say so.

@ianyfan @nmschulte I created this early pull request to let you know I'm working on it. Just in case that we do not double the work. Maybe you could add some other points that should be addressed before this pull request can be merged. I would say keyboard navigation and touch support can be done in a second pr.

emersion commented:
> @ianyfan @nmschulte It seems like #5161 is dead

Yes, this PR has been looking for a new owner for a while. Thanks for taking over!

nmschulte commented:
@FlexW Thank you for doing this.  I would like to support you however I can.

Please see my branch from #5161 work; https://github.com/nmschulte/sway/tree/fix-tray-updates.  It has improved message handling to resolve some the standing issues of #5161; it is based upon current `master/master`, and I rebase it locally about once a week to continue using it.  Since 1.6 / wlroots 1.2/1.3, something has broken for me w/re: non-1.0 DPI scaling and popups, such that most apps no longer work to show menus in this scenario.

This is all I have for now; I will be following closely.  Later I will check your list and my notes and share any new items I've thought about to incorporate.

FlexW commented:
> @FlexW Thank you for doing this. I would like to support you however I can.
> 
> Please see my branch from #5161 work; https://github.com/nmschulte/sway/tree/fix-tray-updates. It has improved message handling to resolve some the standing issues of #5161; it is based upon current `master/master`, and I rebase it locally about once a week to continue using it. Since 1.6 / wlroots 1.2/1.3, something has broken for me w/re: non-1.0 DPI scaling and popups, such that most apps no longer work to show menus in this scenario.
> 
> This is all I have for now; I will be following closely. Later I will check your list and my notes and share any new items I've thought about to incorporate.

Thank you for your support. I'll check your branch out. I think I had the same scaling issue that you described, and I think that issue is resolved in my branch. There was an issue with not specifying the correct size of the buffer when scaling is activated. If you want you can check out my branch and see if the scaling issue is gone for you too.

nmschulte commented:
👍 I hope to dive through the diff later this week.

On Sun, May 2, 2021, 1:41 PM Felix Weilbach ***@***.***>
wrote:

> @FlexW <https://github.com/FlexW> Thank you for doing this. I would like
> to support you however I can.
>
> Please see my branch from #5161 <https://github.com/swaywm/sway/pull/5161>
> work; https://github.com/nmschulte/sway/tree/fix-tray-updates. It has
> improved message handling to resolve some the standing issues of #5161
> <https://github.com/swaywm/sway/pull/5161>; it is based upon current
> master/master, and I rebase it locally about once a week to continue
> using it. Since 1.6 / wlroots 1.2/1.3, something has broken for me w/re:
> non-1.0 DPI scaling and popups, such that most apps no longer work to show
> menus in this scenario.
>
> This is all I have for now; I will be following closely. Later I will
> check your list and my notes and share any new items I've thought about to
> incorporate.
>
> Thank you for your support. I'll check your branch out. I think I had the
> same scaling issue that you described, and I think that issue is resolved
> in my branch. There was an issue with not specifying the correct size of
> the buffer when scaling is activated. If you want you can check out my
> branch and see if the scaling issue is gone for you too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/swaywm/sway/pull/6249#issuecomment-830852893>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBFAT6TYWVL7JLF63C5XS3TLWMENANCNFSM437SF2HQ>
> .
>

Xyene commented:
> something has broken for me w/re: non-1.0 DPI scaling and popups, such that most apps no longer work to show menus in this scenario.

Does https://github.com/swaywm/sway/issues/6147 match what you're describing? If so and you're experiencing it with other apps, would be worth leaving a note in that thread.

nmschulte commented:
> Does #6147 match what you're describing? If so and you're experiencing it with other apps, would be worth leaving a note in that thread.

@Xyene I don't think #6147 describes the issue, but I can try with external displays; experiencing this on a laptop with 4K display, no external monitors.

Zoom does strange things with its menus under Wayland such that the menus popup on the wrong output or don't show at all, depending which output they're parented, on a top-bottom-plus-(rotated-)straddling-sides 4-output DPI==1 stationary desktop setup though.  Also I see #6148 with this same setup.

nmschulte commented:
> Since 1.6 / wlroots 1.2/1.3, something has broken for me w/re: non-1.0 DPI scaling and popups, such that most apps no longer work to show menus in this scenario.

😃 All of the popups seem to be working on this branch (`flexw/tray-dbus-menu`).  `swaybar` crashed during my first smoke test, however.  More to come.

FlexW commented:
> > Since 1.6 / wlroots 1.2/1.3, something has broken for me w/re: non-1.0 DPI scaling and popups, such that most apps no longer work to show menus in this scenario.
> 
>  All of the popups seem to be working on this branch (`flexw/tray-dbus-menu`). `swaybar` crashed during my first smoke test, however. More to come.

Thats great to hear. I'm aware of that crash, it's because I don't handle the hotspots correct fow now. I'll let you know when I fixed it.

Commit:
Tray: Implement dbusmenu

Co-authored-by: Ian Fan <ianfan0@gmail.com>
Co-authored-by: Nathan Schulte <nmschulte@gmail.com>

Signed-off-by: Felix Weilbach <felix.weilbach@t-online.de>

nmschulte commented:
@FlexW after merging `master` and dealing with `wayland-protocol` version issues, I've managed to build this branch and use it for about a week.  I haven't yet had time to review the code changes, however.
I do see a few bugs and `swaybar` crashes still, related to sub-menu interactions.  I'll investigate and follow up when I can.

TimB87 commented:
uneducated question: is there a possibility that this might work with `basu` as well?

emersion commented:
It should, yes.

FlexW commented:
@nmschulte Finally I found some time to work on it. Should now work better. So far I couldn't crash it anymore and it works as I would expect it with all my tray icons (nmapplet, nextcloud, udiskie, blueman, gammastep-indicator). I would appreciate it if you could do some tests too (I rebased against the latest master). Maybe you also know some more applications with which I can test.

FlexW commented:
@ianyfan @emersion One more question: It seems like the `wl_pointer_button`, `wl_pointer_motion` event and dbus events etc run in parallel. How would you advise avoiding race conditions while operating on the `swaybar_dbusmenu` structure?  Mutexes?

emersion commented:
They don't run "in parallel", there's only one thread. The event loop will dispatch D-Bus events and Wayland events one after the other. No need for any kind of locking or race condition prevention.

FlexW commented:
@emersion Thank you for clarifying. Will a [call to](https://github.com/swaywm/sway/pull/6249/files#diff-6665855012dba4194fd662f8add2b51e9cd093589127c1ff3b39c7e86c8e697fR626) `wl_display_roundtrip()` lead to dispatching events? So if I call it in a function I will after `wl_display_roundtrip()` returns have code running in parallel?

nmschulte commented:
I finally managed to build this (caved and built `libdrm`, `wayland-protocols`); sorry my report is shallow.

I managed to crash `swaybar` playing with `udiskie` again; I only managed it once and have no logs.

Both `udiskie` and `nm-applet` menus seem to flicker more than I'd like to see as a user.

Sub-menus are placed a bit off for me (2x scale), and there's an odd looking "bar"/line/border at the bottom of the sub-menus.

![image](https://user-images.githubusercontent.com/8540239/127811728-3a0a718c-4527-424a-a8e5-9b9deb03d910.png)

Finally, and I'm not confident in my sway-term/understanding here, there's still something odd with focus, popups, and xwayland vs xdg_shell (backends?).
I was unable to use my key binding to launch `dmenu`; `gimp`'s menubar would not... popup, even though the interaction feedback seemed to work.  Just prior, I was using `screenshot.sh` (`dmenu`) to take a shot of the open popup(s) from `swaybar` (which works great/is possible as long as I don't have to click), which uses `mako` to notify the success (popup?).
In the end, I was able to rectify things after mashing the keyboard, switching workspaces, clicking / changing focus, etc. without restarting `sway`. :shrug: I've spoke about similar things in the original PR.

FlexW commented:
I would like to get this pr reviewed, because I think it works well enough and provides everything to do basic work with tray menus. There are some things that could be better, but I think it is usable. I can not crash it anymore, `udiskie` does just sometimes blink for me one time when I open the menu. Submenus work also fine.

Maybe one of the maintainers can help with getting the last flickering and focus problems away.

nmschulte commented:
I started reviewing this last night, FYI.  I've found some quirks but they're minor.  It seems like this still includes Ian's (and perhaps some of my) work; I wish you hadn't squashed all of the history is all.

I'm trying to carry all of my effort/discovery forward through your changeset before I'm happy, but so far the diff looks less involved than I had thought.  More to come...

emersion commented:
Please use `Co-authored-by` to credit the appropriate authors when squashing multiple commits together. Squashing commits to have a nice history is good, but it's also important to retain authorship of the work.

@nmschulte Reviews definitely welcome! I'll try to have a look at some point too.

emersion commented:
It seems like this needs a rebase against master.

FlexW commented:
Thank you for reviewing this. I will add the authors. The reason why this is only one commit is because I started a new branch. I wasn't able to transition from @ianyfan approach to load everything as lazy as possible (which made it difficult to make submenus work) to my approach that loads all submenus upfront. A lot of the dbus message handling and a big part of the drawing was a copypaste from @ianyfan PR. I will note that in the commit message.

nmschulte commented:
@FlexW, try this in the commit message:

```
Co-authored-by: Ian Fan <ianfan0@gmail.com>
Co-authored-by: Nathan Schulte <nmschulte@gmail.com>
```

nmschulte commented:
I haven't yet chased everything from my end but here's a video showing a few quirks:
1) the already noted flicker/re-painting
2) cannot interact with other tray items after opening a the menu for an item and never giving it focus; you must first give it focus, or close it by clicking outside the menu and the bar
- a user should be able to open a different tray item menu by simply doing so (right click/`Activate` on it); no extra steps; any open menu(s) automatically close; should not have to interact with a menu in order for it to "be closable
3) scrolling (vertically, with my touchpad]) on the open menu of `nm-applet` is a sure-fire way to crash `swaybar`:
```
...
00:00:07.220 [swaybar/tray/dbusmenu.c:1161] :1.692/org/ayatana/NotificationItem/nm_applet/Menu hovered id 1096
00:00:07.389 [swaybar/tray/dbusmenu.c:1161] :1.692/org/ayatana/NotificationItem/nm_applet/Menu hovered id 1096
00:00:07.671 [swaybar/input.c:306] wl_pointer_axis:axis with no active output
Aborted
```

It's not all menus but seems specific to ones that are dynamic/update, and it seems I must be hovering the menu while scrolling.  As well, scrolling on the menu causes my workspaces to change: just like scrolling on the bar itself.  I don't think this should occur.

nmschulte commented:
d3768e4 and 44f947a seem effective as they claim; I haven't caused a crash yet.

there's a new symptom now though: scrolling events/signals seem to just queue up and now take effect when the cursor enter/leaves (I think that's technically accurate).

FlexW commented:
> [d3768e4](https://github.com/swaywm/sway/commit/d3768e479c3f138aaa52ef7ef5994547a13df1c3) and [44f947a](https://github.com/swaywm/sway/commit/44f947aa9bb1cf36ef384f64867fcdfc89258573) seem effective as they claim; I haven't caused a crash yet.
> 
> there's a new symptom now though: scrolling events/signals seem to just queue up and now take effect when the cursor enter/leaves (I think that's technically accurate).

fixed :) thanks for all your testing.

nmschulte commented:
> fixed
indeed so

a few more quirks:
![quirky-swaybar](https://user-images.githubusercontent.com/8540239/133664727-5d241a31-8a56-4c06-81da-1b31d355c521.gif)

- an action causing an open menu to close should still take the action (e.g. like Firefox and others do, see below)
- an (e.g. scroll; non-"activation" class?) action on the same item as an open menu should not close the menu _(see next para; this I suspect is due to repainting issue)_
- I'm not sure which popup/window wins, but whatever is going on w/re: this is rather crude (menu vs notify)

I'm guessing these won't be fixed (and therefor, won't be fixed... ever), but I think they're defects:
- the state of the menu is lost during repaint: hovered items no longer appear hovered (and there-in: open sub-menus do not re-open)
- flicker due to repaint (probably this and last are one-and-the-same)

![firefox-handling](https://user-images.githubusercontent.com/8540239/133664791-19a68742-f506-4334-88a4-9c6a871b1e1b.gif)

wundrweapon commented:
What's the quickest way for me to run this on my own machine? Doesn't sound like this is getting merged any time soon but I'd like to give it a whirl without risking my current setup. Trying to build fails because it requires wlroots 0.15+ but that isn't a version that exists. I'm new to Wayland; please forgive the stupid question

nmschulte commented:
> What's the quickest way for me to run this on my own machine?

I have a script that builds libseat, wlroots, and sway, with wlroots and libseat as "subprojects."  I sometimes have to also build wayland-protocols and deal with some other lower-level hurdles, but those are infrequent and tend to resolve as my distro "catches up" upstream.  I'll share it here for you, but note it is raw and won't work out of the box, and I won't be able to support it./

https://gist.github.com/nmschulte/66a0b798dded3947695ade6bbd8189f5

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+
+static int handle_item_activation_requested(sd_bus_message *msg, void *data,
+		sd_bus_error *error) {
+	return 0; // TODO
```
Be more explicit about the todo for the next one picking it up please.

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+	close_menu(menu);
+}
+
+static void free_items(struct swaybar_dbusmenu_menu *menu) {
```
`items` and `child_menus` should be freed as well with `list_free`

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+
+static struct swaybar_dbusmenu_menu_item *find_item(
+		struct swaybar_dbusmenu_menu *menu, int item_id) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+
+struct swaybar_dbusmenu_menu *find_menu_id(struct swaybar_dbusmenu_menu *menu,
+		int id) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+	if (menu->item_id == id) {
+		return menu;
+	}
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+	}
+
+	if (menu->child_menus && menu->child_menus->length > 0) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+
+static void swaybar_dbusmenu_draw_menu(struct swaybar_dbusmenu_menu *menu,
+		int id, bool open) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+
+	if (dbusmenu_surface->width != surface_width ||
+			dbusmenu_surface->height != surface_height) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+}
+
+static void swaybar_dbusmenu_draw(struct swaybar_dbusmenu *dbusmenu, int id) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+}
+
+static cairo_surface_t *read_png(const void *data, size_t data_size) {
+	struct png_stream *png_stream = malloc(sizeof(struct png_stream));
```
Check if alloc fails

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+}
+
+static void open_menu_id(struct swaybar_dbusmenu *dbusmenu, int menu_id) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+			log_value = toggle_type;
+		} else if (strcmp(key, "toggle-state") == 0) {
+			sd_bus_message_read(msg, "v", "i", &item->toggle_state);
+			log_value = item->toggle_state == 0 ? "off"
```
Style: tab issues. Should be `on` if 1 or over.

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+			char *children_display;
+			sd_bus_message_read(msg, "v", "s", &children_display);
+			if (strcmp(children_display, "submenu") == 0) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+
+				struct swaybar_dbusmenu_menu *submenu;
+				if (item->id != 0) {
+					submenu = calloc(1, sizeof(struct swaybar_dbusmenu_menu));
```
Is this properly freed?

FlexW replied:
Thanks for the hint. I thought it wasn't.

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+	menu->dbusmenu = dbusmenu;
+	int ret = 0;
+	while (!sd_bus_message_at_end(msg, 1)) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+}
+
+static void swaybar_dbusmenu_get_layout(struct swaybar_dbusmenu *menu, int id) {
+	struct swaybar_sni_slot *slot = calloc(1, sizeof(struct swaybar_sni_slot));
```
Check if alloc fails.

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+	struct swaybar_dbusmenu *menu = data;
+	if (strcmp(interface, xdg_wm_base_interface.name) == 0) {
+		menu->wm_base =
+				wl_registry_bind(registry, name, &xdg_wm_base_interface, 1);
```
Style: 1 tab

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+		wl_list_insert(&menu->sni->slots, &slot->link);
+	} else {
+		sway_log(SWAY_ERROR, "%s%s failed to get IconThemePath: %s",
+				menu->sni->service, menu->sni->menu, strerror(-ret));
```
Style: 1 tab

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+}
+
+static const struct wl_registry_listener registry_listener = {
+		.global = handle_global,
+		.global_remove = handle_global_remove,
```
Style: 1 tab

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+		struct swaybar_output *output, struct swaybar_seat *seat,
+		uint32_t serial, int x, int y) {
+	struct swaybar_dbusmenu *dbusmenu =
+			calloc(1, sizeof(struct swaybar_dbusmenu));
```
Style: 1 tab

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+static void pointer_motion_process_item(
+		struct swaybar_dbusmenu_menu *focused_menu,
+		struct swaybar_dbusmenu_menu_item *item, struct swaybar_seat *seat) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+
+static void open_close_child_menu(struct swaybar_dbusmenu_menu *menu,
+		struct swaybar_dbusmenu_menu_item *item, int x, int y) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+bool dbusmenu_pointer_motion(struct swaybar_seat *seat,
+		struct wl_pointer *wl_pointer, uint32_t time_, wl_fixed_t surface_x,
+		wl_fixed_t surface_y) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+
+static struct swaybar_dbusmenu_menu *dbusmenu_menu_find_menu_surface(
+		struct swaybar_dbusmenu_menu *menu, struct wl_surface *surface) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+static bool dbusmenu_pointer_button_left_process_item(
+		struct swaybar_dbusmenu *dbusmenu,
+		struct swaybar_dbusmenu_menu_item *item, struct swaybar_seat *seat) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1373 @@
+
+bool dbusmenu_pointer_button(void *data, struct wl_pointer *wl_pointer,
+		uint32_t serial, uint32_t time_, uint32_t button, uint32_t state) {
+
```
Style: extra line

bl4ckb0ne commented on swaybar/tray/item.c:
```diff
@@ -333,8 +336,14 @@ void destroy_sni(struct swaybar_sni *sni) {
 
-static void handle_click(struct swaybar_sni *sni, int x, int y,
-		uint32_t button, int delta) {
+static void handle_click(struct swaybar_sni *sni,
+		struct swaybar_output *output,
+		struct swaybar_seat *seat,
+		uint32_t serial,
+		int x,
+		int y,
+		uint32_t button,
+		int delta) {
```
Style: should be back to what it was

bl4ckb0ne commented on swaybar/tray/item.c:
```diff
@@ -384,8 +397,14 @@ static int cmp_sni_id(const void *item, const void *cmp_to) {
 static enum hotspot_event_handling icon_hotspot_callback(
-		struct swaybar_output *output, struct swaybar_hotspot *hotspot,
-		double x, double y, uint32_t button, void *data) {
+		struct swaybar_output *output,
+		struct swaybar_hotspot *hotspot,
+		struct swaybar_seat *seat,
+		uint32_t serial,
+		double x,
+		double y,
+		uint32_t button,
+		void *data) {
```
Style: should be back to what it was

bl4ckb0ne commented on swaybar/tray/item.c:
```diff
@@ -401,7 +420,14 @@ static enum hotspot_event_handling icon_hotspot_callback(
 
 		sway_log(SWAY_DEBUG, "Guessing click position at (%d, %d)", global_x, global_y);
-		handle_click(sni, global_x, global_y, button, 1); // TODO get delta from event
+		handle_click(sni,
+				output,
+				seat,
+				serial,
+				global_x,
+				global_y,
+				button,
+				1); // TODO get delta from event
```
Style: should be back to what it was

bl4ckb0ne reviewed (commented):
I picked up mostly style issues, there's more than pointed out but since it's a pretty big patch I must've missed some.

Havent thoroughly check memory leaks too, might be worth to run it with asan/valgrind if possible to hunt them all down.

I am not familiar with sway's codebase nor the DBus api, we'll need another set of eyes before this can be merged.

Thanks for the patch!

FlexW commented:
@bl4ckb0ne Thanks for reviewing. It might make sense to change the `.clang-format` file to account for this style because the formating in this pr was all done by `clang-format` automatically.

emersion commented:
The `.clang-format` file isn't accurate. clang-format can't describe our style. We should delete it.

GreyXor commented:
@FlexW thanks I will try it

nmschulte commented:
Note: I am able to crash swaybar routinely by simply trying to open Zoom's menu.  I haven't had a chance to diagnose due to the environment, but it is easy for me to reproduce; some day I'll recreate it if I'm not beaten to it.

ianyfan commented:
Thanks for taking this over. How did you manage to fix the udiskie behaviour? I recall that was where I ended up getting stuck.

GreyXor commented:
Interesting discussions on a new standard that may change tray icons https://gitlab.freedesktop.org/xdg/xdg-specs/-/issues/84

michaelweiser commented:
I can reproduce crashes with `blueman-applet` when opening its context menu. It does open and render when right-clicked but then freezes for about ten seconds. After that it closes and quickly thereafter (~0.1s) swaybar crashes. Here's the gdb backtrace and a quick fix check for a null pointer: https://github.com/michaelweiser/sway/commit/841fe081f08c90c73cd9a2c7aa7ed6f2fad03fd3. I've not yet looked further for the root cause. But I remember from my own python-based implementation of StatusNotifierHost/Watcher that blueman-applet was flooding it with duplicate registrations. Maybe something similar is going on here which causes the dbusmenu to be continuously de- and reconstructed?

goto-engineering commented:
This looks really promising! Is there still anything blocking a merge?

wundrweapon commented:
> How did you manage to fix the udiskie behaviour?

Looks like udiskie had some updates after v2.2.0 that addressed either Wayland and Sway specifically, so the issues you had may have actually been on udiskie's end the whole time.

> Is there still anything blocking a merge?

At least one applet (Blueman) seems to have consistent trouble, though I cannot confirm. Also, FlexW is probably too busy with Nextcloud and his own projects that this is on the backburner at best. My guess is that either someone else will have to pick up where he left off, or the Sway guys will have to decide "something mostly working is better than nothing at all" and merge anyway. I may be reading a criminal amount into lines that barely exist, so take that guess with a big fat heap of salt.

tinywrkb commented:
> At least one applet (Blueman) seems to have consistent trouble, though I cannot confirm.

I don't think that blueman is a good example, that's the most buggy application I've encountered that uses dbus name for its agent's singleton.  
I think that most likely whatever issues you hit with it are related to its dbus name acquiring for the agent's singleton, not SNI.

FlexW commented:
> I can reproduce crashes with `blueman-applet` when opening its context menu. It does open and render when right-clicked but then freezes for about ten seconds. After that it closes and quickly thereafter (~0.1s) swaybar crashes. Here's the gdb backtrace and a quick fix check for a null pointer: [michaelweiser@841fe08](https://github.com/michaelweiser/sway/commit/841fe081f08c90c73cd9a2c7aa7ed6f2fad03fd3). I've not yet looked further for the root cause. But I remember from my own python-based implementation of StatusNotifierHost/Watcher that blueman-applet was flooding it with duplicate registrations. Maybe something similar is going on here which causes the dbusmenu to be continuously de- and reconstructed?

I fixed that issue. The problem was mainly that blueman sets the `needUpdate` after every `AboutToShow` notification. This led to an endless loop. I will update the PR tomorrow. It's late here now;)

> Note: I am able to crash swaybar routinely by simply trying to open Zoom's menu. I haven't had a chance to diagnose due to the environment, but it is easy for me to reproduce; some day I'll recreate it if I'm not beaten to it.

I've tested the application with my newest changes and couldn't find any issues or crashes. Maybe I fixed the issue by accident. You may give it a try when I've updated the PR.

FlexW commented:
PR is updated.  @bl4ckb0ne I think I fixed all your comments. I've run the changes also with an ASAN build. Could not find any leaks, but I think it would not cause harm if someone else can test it as well;)

michaelweiser commented:
I've been running this latest commit without any additional changes or fixes for all of today with nm-applet, blueman-tray and remmina without any problems. According to "no news are good news", if I don't report back it continues to work without problems in daily use.

DemiMarie commented:
Is this designed to be secure in the face of malicious input?  I ask because Qubes OS will be switching from XEmbed to SNI, and untrusted guests will be allowed to use SNI.  There will be some sanitization (such as limits on recursion depth and replacement of certain dangerous code points), but I am still worried about security.

Nefsen402 commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+}
+
+static int get_icon_theme_path_callback(sd_bus_message *msg, void *data,
+                                        sd_bus_error *error) {
```
Please follow the standard indentation guidelines

Nefsen402 commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+}
+
+static void handle_global_remove(void *data, struct wl_registry *registry,
+                                 uint32_t name) {
```
Ditto

Nefsen402 commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+};
+
+struct swaybar_dbusmenu *swaybar_dbusmenu_create(struct swaybar_sni *sni,
+                                                 struct swaybar_output *output,
```
Ditto...

Nefsen402 commented on swaybar/meson.build:
```diff
@@ -3,7 +3,8 @@ tray_files = have_tray ? [
 	'tray/tray.c',
-	'tray/watcher.c'
+	'tray/watcher.c',
+    'tray/dbusmenu.c'
```
Tabs not spaces.

Nefsen402 commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+	return NULL;
+}
+
+static void close_menus_id(struct swaybar_dbusmenu_menu *menu, int item_id) {
```
I'd prefer to call this close_menus_by_id

Nefsen402 commented on swaybar/input.c:
```diff
@@ -112,19 +116,38 @@ static void wl_pointer_enter(void *data, struct wl_pointer *wl_pointer,
+#if HAVE_TRAY
 	struct swaybar_seat *seat = data;
 	seat->pointer.x = wl_fixed_to_double(surface_x);
 	seat->pointer.y = wl_fixed_to_double(surface_y);
```
HAVE_TRAY shouldn't be guarding this code. Originally noticed by @kennylevinsen on IRC.

Nefsen402 reviewed (changes requested):
I'm not a fan of how this is implemented. We're making a wayland compositor double up as a wayland client which feels really dirty. I feel like this being its own client application that responds to swayipc would be a lot better.

nmschulte commented:
> I'm not a fan of how this is implemented. We're making a wayland compositor double up as a wayland client which feels really dirty. I feel like this being its own client application that responds to swayipc would be a lot better.

I've been out of the loop since my initial efforts and some of the design changes since, but `swaybar` is its own client application, isn't it?  I'm ignorant: to which aspect are you not a fan of, exactly?  I think the code needs lots of improvement to be a real solution to the problem: refactor/renames and some design/flow changes, avoiding repaints due to menu updates etc., likely requiring a semi-heavy analysis of how tray clients communicate as they do to ensure a seamless UX as I desire: not something I'm interested in doing at the moment.  I do appreciate the efforts that are being taken: my comments are not meant to dissuade.

Nefsen402 commented:
> > I'm not a fan of how this is implemented. We're making a wayland compositor double up as a wayland client which feels really dirty. I feel like this being its own client application that responds to swayipc would be a lot better.
> 
> I've been out of the loop since my initial efforts and some of the design changes since, but `swaybar` is its own client application, isn't it? I'm ignorant: to which aspect are you not a fan of, exactly? I think the code needs lots of improvement to be a real solution to the problem: refactor/renames and some design/flow changes, avoiding repaints due to menu updates etc., likely requiring a semi-heavy analysis of how tray clients communicate as they do to ensure a seamless UX as I desire: not something I'm interested in doing at the moment. I do appreciate the efforts that are being taken: my comments are not meant to dissuade.

Yeah, I was assuming this was baked in with the sway binary since we're working in well, the sway repository. Got my wires crossed. Eventually, I'd like to see swaybar moved into its own repo like swaybg. For now, this is fine I guess.

kennylevinsen commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+			}
+			int i = 0;
+			for (char *c = label; *c; ++c) {
+				if (*c == '_' && !*++c) {
```
Why do we skip underscores here? Without, this would have been a simple `item->label = strdup(label);`.

FlexW replied:
From the dbusmenu protocol documentation:
```
              <tr>
                        <td>label</td>
                        <td>string</td>
                        <td>Text of the item, except that:
                        -# two consecutive underscore characters "__" are displayed as a
                        single underscore,
                        -# any remaining underscore characters are not displayed at all,
                        -# the first of those remaining underscore characters (unless it is
                        the last character in the string) indicates that the following
                        character is the access key.
                        </td>
                        <td>""</td>
                </tr>
```

nmschulte replied:
I believe a single underscore should be retained, however: only the underscores following two underscores should be elided -- that's how I read that spec.

nmschulte replied:
Add logic to display the escaped underscores (`__`) as a single underscore?  The current logic just throws away all underscores as I read it, which isn't to spec.

FlexW replied:
I haven't written the logic it's from @ianyfan, but I believe it is correct. The `!*++c` skips the first '_'. Then if there is a second '_' it will be added to the label.

FlexW replied:
`!*++c` means: Go to the next character. Dereference it and if it's 0 then the statement becomes true. That means doing nothing and just adding the next character to the label if we haven't reached the end of the string. In case, we reached the end of the string the loop can be terminated.

Honsetly, I'm not a fan of this code. If we have such a discussion here it should probably be simplified to make it easier to reason about.

nmschulte replied:
Ah ... sorry, too much JavaScript on my brain, not really an excuse though: I flat-out missed the inversion.
I think still though this is ignoring the spec; `____` will render as `__`, and we ignore the accelerator keys etc. -- I don't think this should hold up this effort, but there are lots of little things like this that could be improved.

kennylevinsen commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+			}
+			log_value = children_display;
+		} else {
+			// Ignored: shortcut, disposition, disposition
```
Disposition is written twice.

kennylevinsen commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+	dbusmenu->seat = seat;
+	dbusmenu->serial = serial;
+	dbusmenu->x = seat->pointer.x;
+	dbusmenu->y = seat->pointer.y;
```
I imagine this was supposed to use the `x` and `y` arguments? Otherwise, remove them.

kennylevinsen commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+		return NULL;
+	}
+
+	sni->tray->menu = dbusmenu;
```
If we assign this here, we should check if it's non-NULL first. In that case, maybe `swaybar_dbusmenu_open` to indicate that it's not just a constructor.

kennylevinsen commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+	}
+}
+
+static void close_child_menus_outside_pointer(struct swaybar_dbusmenu_menu *menu,
```
Maybe "unfocused menus" rather than "menus outside pointer"?

kennylevinsen commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+			time(NULL));
+
+		if (!tray->menu->drawing) {
+			swaybar_dbusmenu_destroy(tray->menu);
```
Why do we destroy the menu if we are not drawing it?

FlexW replied:
The menu was only allowed to be destroyed if no drawing was ongoing in parallel, I removed the drawing variable as it now not necessary anymore.

kennylevinsen commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+	struct swaybar_tray *tray = menu->dbusmenu->sni->tray;
+	menu = find_menu_id(menu->dbusmenu->menu, id);
+	if (!menu) {
+		menu->dbusmenu->drawing = false;
```
`NULL` pointer dereference, you've just overwritten `menu`.

kennylevinsen commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+		xdg_surface_add_listener(xdg_surface, &xdg_surface_listener, NULL);
+		wl_surface_commit(surface);
+
+		wl_display_roundtrip(bar->display);
```
We do not need to do a roundtrip, you can destroy the positioner immediately after use as the messages are queued.

I believe this roundtrip is why the "drawing" state was added as well, right? If so, that can be removed as well as it makes the code rather confusing.

FlexW replied:
Correct, that makes it easier. Thanks.

kennylevinsen replied:
This is not solved with a roundtrip, it's solved by not rendering anything until you have received and acid the first configure event for the surface, sent when you assigned the xdg popup role.

FlexW replied:
Thanks, I got rid of the roundtrip.

kennylevinsen commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+			dbusmenu_surface->height != surface_height) {
+		if (dbusmenu_surface->surface) {
+			wl_surface_destroy(dbusmenu_surface->surface);
+			dbusmenu_surface->surface = NULL;
```
What about `xdg_surface_destroy`?

kennylevinsen reviewed (changes requested):
Thanks for the work so far!

I have not checked the rendering logic or tested this, but I did a cursory review and find some issues that need to be fixed and improvements that should be made before we can consider merging this.

FlexW commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+			if (item->icon_name) {
+				free(item->icon_name);
+			}
+			free(item);
```
good catch. thanks.

FlexW commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1381 @@
+	struct swaybar_dbusmenu *dbusmenu;
+	struct swaybar_dbusmenu_surface *surface;
+	list_t *items;       // struct swaybar_dbusmenu_menu_item
+	list_t *child_menus; // struct swaybar_dbusmenu_menu
```
Probably can be done without `child_menus` and `parent_menu`. I will try to do that, however that will take a bit of time and testing.

FlexW replied:
I have removed `child_menus` and `parent_menu`.

FlexW commented:
I've fixed most review comments. What's left to do:
- [ ] Remove `child_menus` if possible
- [ ] Remove `parent_menu` if possible
- [ ] Remove `wl_display_roundtrip()` if possible (waiting for feedback)

ek9 commented:
would love to see this merged - can I help with anything regarding the issue?

joanbm commented:
This needs to be rebased for Sway 1.8-rc1, there are some conflicts due to 94b69acf0d7b26ee5af2172300cb18473508da76 and 53f9dbd424dc173a85c9f4cd30802259d38b1ef4, mostly trivial, related to function arguments. I attach my rebased version in case it helps save some work:
[pr-6249-dbus-menus-rebase-1.8-rc1.patch.zip](https://github.com/swaywm/sway/files/10147631/pr-6249-dbus-menus-rebase-1.8-rc1.patch.zip)

---

On the other hand, at some point, the line in `swaybar_dbusmenu_open` that read `dbusmenu->x = seat->pointer.x;`  was changed to `dbusmenu->x = x;` (ditto for `y`). This breaks the positioning logic, it only seems to work with an output positioned at (0, 0) and with the bar positioned at the top. Reverting the change fixes it on all cases I tested (i.e. with secondary monitors, bar at bottom or top, scale, gaps):

```diff
diff --git a/swaybar/tray/dbusmenu.c b/swaybar/tray/dbusmenu.c
index b8b969d0..a1cee50b 100644
--- a/swaybar/tray/dbusmenu.c
+++ b/swaybar/tray/dbusmenu.c
@@ -1066,8 +1066,8 @@ void swaybar_dbusmenu_open(struct swaybar_sni *sni,
 	dbusmenu->output = output;
 	dbusmenu->seat = seat;
 	dbusmenu->serial = serial;
-	dbusmenu->x = x;
-	dbusmenu->y = y;
+	dbusmenu->x = seat->pointer.x;
+	dbusmenu->y = seat->pointer.y;
 	dbusmenu->bar = output->bar;
 
 	swaybar_dbusmenu_setup(dbusmenu);
```

FlexW commented:
@joanbm Thanks for the diff. I've updated the PR with it.

nmschulte commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1368 @@
+		swaybar_dbusmenu_get_layout(sni->tray->menu, menu_id);
+	}
+
+    swaybar_dbusmenu_draw(sni->tray->menu, menu_id);
```
nit: indentation (spaces vs tabs)

FlexW replied:
Thanks, I fixed some other whitespace issues as well.

nmschulte commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1368 @@
+		struct swaybar_dbusmenu_menu_item *item = menu->items->items[i];
+		struct swaybar_dbusmenu_menu *child_menu = item->submenu;
+		if (child_menu && child_menu->item_id == item_id && child_menu->item_id != 0) {
+            close_menus(child_menu);
```
nit: indentation (spaces vs tabs)

nmschulte commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1368 @@
+		int y = seat->pointer.y * scale;
+		if (item->submenu && item->submenu->item_id != 0 
+				&& !is_in_hotspot(&item->hotspot, x, y)) {
+		  close_menus_by_id(menu, item->id);
```
nit: indentation (spaces vs tabs)

nmschulte commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1368 @@
+	if (is_in_hotspot(&item->hotspot, seat->pointer.x * scale,
+			seat->pointer.y * scale)) {
+		if (!item->enabled || item->is_separator) {
+		  return false;
```
nit: indentation

FlexW commented:
@kennylevinsen I resolved all review comments. I appreciate another review round.

mattfbacon commented:
FYI, you can squash when the PR is merged so you don't have to force-push and lose the history.

bim9262 commented:
I've been using this commit for a few weeks now at it's working great, but just noticed that the tray menus only display on my leftmost screen, but not on my right screen.

TimB87 commented:
I noticed that right clicking the QOwnNotes icon (among others that I forgot) have the ability to crash the bar (I use this with i3status-rust).
I'll try to get a trace why this happens.

michaelweiser commented:
> noticed that the tray menus only display on my leftmost screen, but not on my right screen.

Same here. I've played around with it a bit and found that the menu only shows on the display rooted at coordinates 0 0. So with the following in sway config it shows only on eDP-1:
```
output eDP-1 position 0 0
output DP-1 position -3840 -1080
```
With this it shows on DP-1 and not on eDP-1:
```
output DP-1 position 0 0
output eDP-1 position 3840 0
```
It doesn't matter whether the coordinates are negative or not. With this it still only shows on eDP-1:
```
output eDP-1 position 0 0
output DP-1 position 1920 0
```
With this config it shows on neither display:
```
output DP-1 position 1920 10
output eDP-1 position 5760 10
```
With this config it shows on DP-1 but offset to the right by what could well be 300 pixels:
```
output DP-1 position 300 0
output eDP-1 position 4140 00
```
With this config swaybar crashes when clicking on any tray icon on any display:
```
output DP-1 position -300 0
output eDP-1 position 3540 00
```
With "shows" I mean that the icons are there on all displays but only on one display clicking on them actually shows the menu. On the other nothing happens when clicking the icon. The click does seem to reach swaybar though since in both cases it logs:
```
/usr/libexec/gdm-wayland-session[472736]: 00:06:30.495 [ERROR] [swaybar/tray/dbusmenu.c:1060] :1.21/org/blueman/sni/menu failed to read IconThemePath: Invalid argument
```
(In this case blueman tray just as an indicator that the action is received and the actual error very likely totally unrelated to showing the menu or not as it happened with older commits as well which showed the menu on both displays.)

joanbm commented:
@michaelweiser @bim9262 This seems the same as the positioning issue I mentioned on  https://github.com/swaywm/sway/pull/6249#issuecomment-1336291773

michaelweiser commented:
@joanbm: Duh, yes. I can confirm that the change from [your comment ](https://github.com/swaywm/sway/pull/6249#issuecomment-1336291773) fixes the issue. Thanks!

FlexW commented:
I've included the fix from @joanbm. Thanks for all the testing. I don't use that PR myself. I switched a long time ago to Waybar. However, I'm still interested in getting this merged.

lbatalha commented:
I have replicated @TimB87 crash with `QOwnNotes`. It appears I can only make it crash on right click once after the program is first started, when I restart swaybar everything works from then on.
To reproduce:
- Start sway
- Start QOwnNote (with tray icon enabled in settings)
- Right click the tray icon - nothing should happen
- Move the mouse to another output - swaybar will crash

Notes:
- Moving focus in the same way using the keyboard does not seem to trigger this, only with the mouse (perhaps there is another method I am missing, but just $mod+5 to workspace #5 on my right monitor works
- My output layout consists of 4 displays as such:
```
3 horizontal displays with a top center one, all effectively scaled to 3840x2160
output "leftdisplay" pos 0 2160 res 3840x2160 scale 1
output "topcenterdisplay" pos 3840 0 res 1920x1080 scale 0.5
output "right display" pos 7680 2160 res 2560x1440 scale 0.666666
output "centerdisplay" pos 3840 2160 res 3840x2160 scale 1
```
- Right clicking the tray in any of the outputs causes this problem
- This only seems to happen the first time the program is launched, after I reload sway the right click menu works indefinitely as far as I can tell

I tried getting a backtrace, and debug logging, but did not seems like much:
```
Jan 23 23:36:16 home sway[2035]: 00:00:51.969 [INFO] [swaybar/tray/host.c:24] Registering Status Notifier Item ':1.87/StatusNotifierItem'
Jan 23 23:36:28 home sway[2035]: Warning: Wayland does not support QWindow::requestActivate() ((null):0, (null))
Jan 23 23:36:28 home sway[2035]: Warning: Wayland does not support QWindow::requestActivate() ((null):0, (null))
Jan 23 23:36:29 home sway[2035]: 00:01:05.087 [ERROR] [swaybar/tray/dbusmenu.c:892] :1.87/NO_DBUSMENU failed to get layout: No such object path '/NO_DBUSMENU'
Jan 23 23:36:29 home sway[2035]: 00:01:05.087 [ERROR] [swaybar/tray/dbusmenu.c:1060] :1.87/NO_DBUSMENU failed to read IconThemePath: Invalid request descriptor
[GTK errors here from things that are in the tray, like transmission-gtk]
Jan 23 23:36:38 home sway[2035]: thread 'main' panicked at 'failed printing to stdout: Broken pipe (os error 32)', library/std/src/io/stdio.rs:1017:9
Jan 23 23:36:38 home sway[2035]: stack backtrace:
Jan 23 23:36:38 home sway[2035]:    0:     0x55660d44b7dc - <unknown>
Jan 23 23:36:38 home sway[2035]:    1:     0x55660d4729ec - <unknown>
Jan 23 23:36:38 home sway[2035]:    2:     0x55660d444fe1 - <unknown>
Jan 23 23:36:38 home sway[2035]:    3:     0x55660d44d425 - <unknown>
Jan 23 23:36:38 home sway[2035]:    4:     0x55660d44d0c8 - <unknown>
Jan 23 23:36:38 home sway[2035]:    5:     0x55660d44d983 - <unknown>
Jan 23 23:36:38 home sway[2035]:    6:     0x55660d44d867 - <unknown>
Jan 23 23:36:38 home sway[2035]:    7:     0x55660d44bc74 - <unknown>
Jan 23 23:36:38 home sway[2035]:    8:     0x55660d44d589 - <unknown>
Jan 23 23:36:38 home sway[2035]:    9:     0x55660d0b36d3 - <unknown>
Jan 23 23:36:38 home sway[2035]:   10:     0x55660d444243 - <unknown>
Jan 23 23:36:38 home sway[2035]:   11:     0x55660d10d155 - <unknown>
Jan 23 23:36:38 home sway[2035]:   12:     0x55660d1451af - <unknown>
Jan 23 23:36:38 home sway[2035]:   13:     0x55660d229ef3 - <unknown>
Jan 23 23:36:38 home sway[2035]:   14:     0x55660d2487f9 - <unknown>
Jan 23 23:36:38 home sway[2035]:   15:     0x55660d43ed3e - <unknown>
Jan 23 23:36:38 home sway[2035]:   16:     0x55660d149812 - main
Jan 23 23:36:38 home sway[2035]:   17:     0x7fb917f22290 - <unknown>
Jan 23 23:36:38 home sway[2035]:   18:     0x7fb917f2234a - __libc_start_main
Jan 23 23:36:38 home sway[2035]:   19:     0x55660d0b39c5 - <unknown>
Jan 23 23:36:38 home sway[2035]:   20:                0x0 - <unknown>
Jan 23 23:36:38 home sway[2035]: 00:01:15.093 [ERROR] [sway/ipc-server.c:206] IPC Client socket error, removing client
```

The `QWindow::requestActivate()` Qt warnings are probably unrelated, and a Qt bug, since I get them in other Qt programs, they appear each time the program appears when I left click the tray icon. The first log message was when I launched the program the first time.
The `dbusmenu.c` error on line `1060` happens each time regardless on every subsequent right click, even after I have reloaded sway with a new swaybar, and no crashes happen.

TimB87 commented:
@FlexW Yes I am.
@lbatalha Thanks for adding the info, I haven't had the time yet!

ianyfan commented on swaybar/tray/dbusmenu.c:
```diff
@@ -0,0 +1,1367 @@
+		struct wl_pointer *wl_pointer, uint32_t time_, wl_fixed_t surface_x,
+		wl_fixed_t surface_y) {
+	struct swaybar_tray *tray = seat->bar->tray;
+	struct swaybar_dbusmenu_menu *focused_menu = tray->menu_pointer_focus;
+	if (!(tray && tray->menu && focused_menu)) {
```
I've started testing this branch out, and I just got a segmentation fault here due to `tray` being NULL, so this `!tray` check may need to occur sooner.

However, the context was that I was running a bar without a tray. I'm not entirely sure why this callback even runs in that case, it would be good to check whether it makes sense to try to disable it.

FlorianFranzen replied:
Could you test if my patch [below](https://github.com/swaywm/sway/pull/6249#issuecomment-1543543643) alleviates the issue? The null check should probably still be fixed, otherwise it serves no purpose.

SebTM commented:
Hey, I was watching this PR frequently and happy about the activity but it seems to stagnate a bit, is there something I / we can support on to get this further? Personally it's an important feature to me :)

FlorianFranzen commented:
I had to fix some uninitialized variables and add some check to not invoke the dbus menu when the tray is disabled and while it works mostly well, I still get some random crashes when interacting with the tray, some inside `dbusmenu_menu_find_menu_surface`, so there is clearly some work left to do. 

<details>
  <summary>You can find my changes below.</summary>

  ```patch
diff --git a/swaybar/input.c b/swaybar/input.c
index 4ee4915991..8487eeb74a 100644
--- a/swaybar/input.c
+++ b/swaybar/input.c
@@ -118,8 +118,9 @@ static void wl_pointer_enter(void *data, struct wl_pointer *wl_pointer,
 	update_cursor(seat);
 
 #if HAVE_TRAY
-	if (dbusmenu_pointer_enter(data, wl_pointer, serial, surface, surface_x,
-		surface_y)) {
+	struct swaybar_config *config = seat->bar->config;
+	if (!config->tray_hidden && dbusmenu_pointer_enter(data, wl_pointer, serial,
+		surface, surface_x, surface_y)) {
 		return;
 	}
 #endif
@@ -127,13 +128,14 @@ static void wl_pointer_enter(void *data, struct wl_pointer *wl_pointer,
 
 static void wl_pointer_leave(void *data, struct wl_pointer *wl_pointer,
 		uint32_t serial, struct wl_surface *surface) {
+	struct swaybar_seat *seat = data;
 #if HAVE_TRAY
-	if (dbusmenu_pointer_leave(data, wl_pointer, serial, surface)) {
+	struct swaybar_config *config = seat->bar->config;
+	if (!config->tray_hidden && dbusmenu_pointer_leave(data, wl_pointer, serial,
+		surface)) {
 		return;
 	}
 #endif
-
-	struct swaybar_seat *seat = data;
 	seat->pointer.current = NULL;
 }
 
@@ -143,7 +145,9 @@ static void wl_pointer_motion(void *data, struct wl_pointer *wl_pointer,
 	seat->pointer.x = wl_fixed_to_double(surface_x);
 	seat->pointer.y = wl_fixed_to_double(surface_y);
 #if HAVE_TRAY
-	if (dbusmenu_pointer_motion(data, wl_pointer, time, surface_x, surface_y)) {
+	struct swaybar_config *config = seat->bar->config;
+	if (!config->tray_hidden && dbusmenu_pointer_motion(data, wl_pointer, time,
+		surface_x, surface_y)) {
 		return;
 	}
 #endif
@@ -185,8 +189,9 @@ static void wl_pointer_button(void *data, struct wl_pointer *wl_pointer,
 		uint32_t serial, uint32_t time, uint32_t button, uint32_t state) {
 	struct swaybar_seat *seat = data;
 #if HAVE_TRAY
-	if (dbusmenu_pointer_button(seat, wl_pointer, serial, time, button,
-				state)) {
+	struct swaybar_config *config = seat->bar->config;
+	if (!config->tray_hidden && dbusmenu_pointer_button(seat, wl_pointer, serial,
+		time, button, state)) {
 		return;
 	}
 #endif
@@ -308,7 +313,8 @@ static void wl_pointer_axis(void *data, struct wl_pointer *wl_pointer,
 	}
 
 #if HAVE_TRAY
-	if (dbusmenu_pointer_axis(data, wl_pointer)) {
+	struct swaybar_config *config = seat->bar->config;
+	if (!config->tray_hidden && dbusmenu_pointer_axis(data, wl_pointer)) {
 		return;
 	}
 #endif
@@ -330,7 +336,8 @@ static void wl_pointer_frame(void *data, struct wl_pointer *wl_pointer) {
 	struct swaybar_output *output = pointer->current;
 
 #if HAVE_TRAY
-	if (dbusmenu_pointer_frame(data, wl_pointer)) {
+	struct swaybar_config *config = seat->bar->config;
+	if (!config->tray_hidden && dbusmenu_pointer_frame(data, wl_pointer)) {
 		return;
 	}
 #endif
diff --git a/swaybar/tray/dbusmenu.c b/swaybar/tray/dbusmenu.c
index 8821cacaec..909730bea5 100644
--- a/swaybar/tray/dbusmenu.c
+++ b/swaybar/tray/dbusmenu.c
@@ -592,7 +592,7 @@ static void swaybar_dbusmenu_draw_menu(struct swaybar_dbusmenu_menu *menu,
 		cairo_surface_destroy(recorder);
 		return;
 	}
-	int surface_x, surface_y, surface_width, surface_height;
+	int surface_x = 0, surface_y = 0, surface_width = 0, surface_height = 0;
 	draw_menu_items(cairo, menu, &surface_x, &surface_y, &surface_width,
 					&surface_height, open);
 
diff --git a/swaybar/tray/tray.c b/swaybar/tray/tray.c
index b0545f4a70..e4a680342b 100644
--- a/swaybar/tray/tray.c
+++ b/swaybar/tray/tray.c
@@ -110,7 +110,7 @@ static int cmp_output(const void *item, const void *cmp_to) {
 
 uint32_t render_tray(cairo_t *cairo, struct swaybar_output *output, double *x) {
 	struct swaybar_config *config = output->bar->config;
-	if (config->tray_outputs) {
+	if (config->tray_outputs && !config->tray_hidden) {
 		if (list_seq_find(config->tray_outputs, cmp_output, output) == -1) {
 			return 0;
 		}
  ```
  
</details>

@FlexW Are you planing to continue your work on this?

DemiMarie commented:
@FlorianFranzen ASAN+UBSAN should help diagnose the crashes.  Afterwards, I recommend fuzzing the D-Bus API to ensure it is robust against e.g. malicious sandboxed flatpak apps.

DemiMarie commented:
It’s quite possible that some of the crashes are due to libdbusmenu being buggy — it is unmaintained for several years now (since 2017, IIUC) and Waybar also has problems with it.

illegalcoding commented:
Any news on this?

NickHu commented:
This patch no longer cleanly applies on newer versions of sway.
I've rebased this on top of sway 1.9 here: https://github.com/NickHu/sway/tree/feature/tray-dbus-menu-1.9, incorporating @FlorianFranzen's changes above.
As far as I am concerned, I have been using this patch for years and never experienced any instability, so even if it's not perfect and there are edge cases I would be happy to see this merged.

Jamim commented:
_Hello @NickHu,_

> I've rebased this on top of sway 1.9 here: [NickHu/sway@`feature`/tray-dbus-menu-1.9](https://github.com/NickHu/sway/tree/feature/tray-dbus-menu-1.9?rgh-link-date=2024-03-02T11%3A35%3A24Z), incorporating @FlorianFranzen's changes above.

Would you mind submitting an alternative PR?

_Thanks in advance!_

DemiMarie commented:
@NickHu would you mind fuzzing this to ensure it is safe and secure against malicious D-Bus clients?

NickHu commented:
No thank you, I don't have either the expertise nor the willingness to undertake any development in sway. Anyone else is free to do whatever they like with my tiny contribution of rebasing.

llyyr commented:
I've been daily driving this for a few days now, and besides one issue that I fixed here (https://github.com/llyyr/sway/commit/84481c26a4c55674da5804bb2619d3575ba405b3), it seems mostly stable now across the clients I use.

fwiw I also rebased it on master here in case there's other people that want to test it out. https://github.com/llyyr/sway/tree/dbus-tray

I don't really know what's needed to push this further besides some style nits I noticed

DemiMarie commented:
> I've been daily driving this for a few days now, and besides one issue that I fixed here ([llyyr@84481c2](https://github.com/llyyr/sway/commit/84481c26a4c55674da5804bb2619d3575ba405b3)), it seems mostly stable now across the clients I use.
> 
> fwiw I also rebased it on master here in case there's other people that want to test it out. https://github.com/llyyr/sway/tree/dbus-tray
> 
> I don't really know what's needed to push this further besides some style nits I noticed

Can you make a PR with all of the fixes?

grazzolini commented:
I have been keeping these patches on my own fork for a while now and I building a version of this package on Arch without any issues. What is preventing this from being merged?

eclairevoyant commented:
The changes won't merge, so a new PR has to be made, or this one needs to be rebased/force-pushed. Additionally someone would need to incorporate the testing mentioned [here](https://github.com/swaywm/sway/pull/6249#issuecomment-2002510008) and the fix mentioned [here](https://github.com/swaywm/sway/pull/6249#issuecomment-2305834806).

grazzolini commented:
> The changes won't merge, so a new PR has to be made, or this one needs to be rebased/force-pushed. Additionally someone would need to incorporate the testing mentioned [here](https://github.com/swaywm/sway/pull/6249#issuecomment-2002510008) and the fix mentioned [here](https://github.com/swaywm/sway/pull/6249#issuecomment-2305834806). So far no one else has submitted a PR to supercede this one.

The [tray-dbus-menu-1.9](https://github.com/grazzolini/sway/tree/tray-dbus-menu-1.9) branch on my fork contains all the patches and fixes from this branch. However, I'm not sure about the testing part.

grazzolini commented:
> > The [tray-dbus-menu-1.9](https://github.com/grazzolini/sway/tree/tray-dbus-menu-1.9) branch on my fork contains all the patches and fixes from this branch.
> 
> Feel free to open a PR then to drive this forward.

Sure, I'll create a new branch, because that one is targeting the latest release, not main.

jackpurrin commented:
I kinda need to use this patch, and I see people using it, but how can I actually start using it?

Dantali0n commented:
We now have at least #8698 #8405 and this one, #6249 should we try to consolidate a bit? because otherwise we might run in to more and more confused users, especially since some of the patches are already used in the wild.

jackpurrin commented:
@GreyXor Yeah I use Arch, so a PKGBUILD would be appreciated!!

grazzolini commented:
I am not using sway at the moment, which is the reason I haven't touched this patch in a while. I was using a fork of sway on my personal account on top of which I applied the patches and then used Arch's sway PKGBUILD to point to this fork as the source.

mattfbacon commented:
@Dantali0n #8698 is a rebase of #8405 which is a cleaned-up and patched version of this PR. So if we were to move forward it would be with #8698.

GreyXor commented:
> @GreyXor Also if you are going to provide a PKGBUILD, do I need to uninstall sway before I make it or will it replace my current sway install?

Hello, it will replace it.
[here is the PKGBUILD](https://gist.github.com/GreyXor/04ceb748b3f670e2a48c8c10b0f600b6)
So basically you can git clone the [sway-git repo](https://aur.archlinux.org/packages/sway-git) replace the PKGBUILD with mine and run `makepkg -si` in the directory. You also need to wait for https://github.com/swaywm/sway/pull/8698 to be rebased from sway master

pd2s commented:
This might be useful here: I'm currently working on a new widget library and as an example I implemented a swaybar with full tray dbusmenu support: https://github.com/pd2s/sw/tree/master/examples/sw_swaybar

autoteelar commented:
does this work yet? i am trying to right click tray icons in sway
